matgen/generate_matrix_he_and_tz.cc

```diff
@@ -0,0 +1,409 @@
+                    int64_t j_start = A.uplo() == Uplo::Upper ? i  : 0;
+                    int64_t j_end   = A.uplo() == Uplo::Upper ? nt : std::min( i+1, nt );
+                    for (int64_t j = j_start; j < j_end; ++j) {
+                        if (A.tileIsLocal( i, j )) {
+                            #pragma omp task slate_omp_default_none shared( A ) \
+                                firstprivate( i, j, ii, zero )
+                            {
+                                A.tileGetForWriting( i, j, LayoutConvert::ColMajor );
+                                auto Aij = A( i, j );
+                                lapack::laset(
+                                    lapack::MatrixType::General,
+                                    1, Aij.nb(), zero, zero,
+                                    &Aij.at( ii, 0 ), Aij.stride() );
+                            }
+                        }
+                    }
+                }
+                i_global += A.tileMb( i );
+            }
+        }
+    }
+
+    A.tileUpdateAllOrigin();
+}
+
+// -----------------------------------------------------------------------------
+/// Generates an m-by-n Hermitian-storage test matrix.
+/// Handles Hermitian matrices.
+/// Diagonal elements of a Hermitian matrix must be real;
+/// their imaginary part must be 0.
+/// @see generate_matrix
+/// @ingroup generate_matrix
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<scalar_t>& A,
+    std::vector< blas::real_type<scalar_t> >& Sigma,
+    slate::Options const& opts)
+{
+    slate::BaseTrapezoidMatrix<scalar_t>& TZ = A;
+    generate_matrix( params, TZ, Sigma, opts );
+
+    // Set diagonal to real.
+    #pragma omp parallel for
+    for (int64_t i = 0; i < A.mt(); ++i) {
+        if (A.tileIsLocal( i, i )) {
+            A.tileGetForWriting( i, i, LayoutConvert::ColMajor );
+            auto T = A( i, i );
+            int64_t mb = T.mb();
+            for (int64_t ii = 0; ii < mb; ++ii) {
+                T.at( ii, ii ) = std::real( T( ii, ii ) );
+            }
+        }
+    }
+    A.tileUpdateAllOrigin();
+}
+
+// -----------------------------------------------------------------------------
+/// Overload without Sigma.
+/// @see generate_matrix()
+/// @ingroup generate_matrix
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::HermitianMatrix<scalar_t>& A,
+    slate::Options const& opts)
+{
+    using real_t = blas::real_type<scalar_t>;
+    std::vector<real_t> dummy;
+    generate_matrix( params, A, dummy, opts );
+}
+
+template <typename scalar_t>
+void generate_matrix(
+    MatrixParams& params,
+    slate::BaseTrapezoidMatrix<scalar_t>& A,
+    slate::Options const& opts)
+{
+    using real_t = blas::real_type<scalar_t>;
+    std::vector<real_t> dummy;
+    generate_matrix( params, A, dummy, opts );
+}
+
+//------------------------------------------------------------------------------
```

Comment:
@mgates3 I added two explicit sections for the instantiations. One for hermitian and one for trapezoid.  I added two to give a bit more documentation. However, if you don't like that let me know and I will go back to a single section for instantiations.